### ISSUE TITLE
Multiple issues - Fleet integrations across multiple spaces confuses Kibana/Fleet

### ISSUE THREAD

**Issue by colin-stubbs** [elastic/integrations#3434]
**Issue 1**

Fleet integrations and policies are not constrained to a space, e.g. all integrations and all policies are all available to any user who has sufficient privileges to see them, in all spaces. It does not matter whether they're in "space1", they will always see all integrations and all policies created in "space1" and in "space2".

This is confusing, and limits the usefulness of spaces, it also bypasses intended RBAC and space access restrictions revealing information to users who should not be able to see integrations in other spaces.

This also means they can access configured integrations and access credentials and other secrets (API keys etc) that they should not be able to access.

Kibana/Fleet should understand and respect the relationship between individual integration instances and the space they have been deployed in when multiple integrations across multiple spaces are used, only showing the relevant integrations installed within the current space as installed. Integrations deployed in other spaces should not be seen by a user working within a space.

**Issue 2**

A direct effect of issue 1.

Integration assets are only deployed in the first space the integration is first deployed in.

It is not possible to redeploy integration assets in other spaces, except by manually copying them between spaces from the "Management" -> "Saved Objects" page.

Or by exporting and reimporting saved object NDJSON files.

Kibana/Fleet should identify missing assets, and permit installation and reinstallation (overwrite) of all Kibana assets for an integration, into any space, if they do not currently exist as a Saved Object in the current space the user is working within.

**Issue 3**

Combination of issue 1 and 2.

Using the Palo Alto Networks Logs integration as an example.

If I have 2 x integration instances, both on v1.6.0 and v2.1.0 has just been released.

The first integration is in space ID "space1" and the second integration is in space ID "space2"

If I use the upgrade function in Kibana/Fleet webUI, it will upgrade the first instance of a deployed integration to v2.1.0, in "space1"

But it will leave the integration in "space2" on v1.6.0 and it will not upgrade it.

At this point there are no options in the webUI to manually upgrade the v1.6.0 integration that has been left hanging, regardless of whether I attempt this from the "default" space, or "space1" or "space2".

There is no documentation available on how to correct this, and no obvious API endpoint to trigger upgrade of the integration.

The only option seems to be to delete the v1.6.0 integration and recreate it.

This also appears to mean that the ingest pipeline for v1.6.0 no longer exists, as the only integration pipeline in existence at this point is the v2.1.0.

I believe this means that logs will be lost by agents assigned the v1.6.0 integration.

**Comment by colin-stubbs** [comment]
This is on all Elastic Cloud deployments we have by the way, all on v8.2.0.

**Comment by colin-stubbs** [comment]
Example with Azure integration,

![Screenshot 2022-05-26 115113](https://user-images.githubusercontent.com/3059577/170399312-836aa9dd-3d12-4ef0-a486-bbf845289687.png)
![Screenshot 2022-05-26 115202](https://user-images.githubusercontent.com/3059577/170399319-ce7c71c5-b1c1-49e8-a590-d6236291db4a.png)

**Comment by colin-stubbs** [comment]
Despite namespaces in the saved object being a list, so it would appear intended to allow existence/visibility of a single saved object in multiple spaces... the list does not appear to be modifiable... the API errors.

```
PUT {{kibana_url}}/s/not-default/api/saved_objects/dashboard/zookeeper-467207a0-231e-11e9-bb66-8baac426dfd4
{
    "attributes": {
        "namespaces": [
            "not-default",
            "default"
        ]
    }
}
```

```
{
    "statusCode": 400,
    "error": "Bad Request",
    "message": "mapping set to strict, dynamic introduction of [namespaces] within [dashboard] is not allowed: strict_dynamic_mapping_exception: [strict_dynamic_mapping_exception] Reason: mapping set to strict, dynamic introduction of [namespaces] within [dashboard] is not allowed"
}
```

**Comment by colin-stubbs** [comment]
Same thing creating a new object from scratch.

**Comment by colin-stubbs** [comment]
Cool... so:

1. Saved object ID's need to be globally unique, as they're not space specific even though the way the UI etc works suggests that they are.
2. Fleet integration management features assume that saved object ID's that come as part of the integration will be the same in the space you're currently viewing them in, but they're not.
3. Saved objects should be sharable between spaces, but they aren't, they're only "copyable"
4. Various other features like Elastic Security also assume that saved objects will exist with a specific object ID that won't in fact exist if multiple spaces are in use.

Neat.

Can someone take this on as a bug fix please? It seems like Elastic Security integrations and use of saved objects hasn't really been thought thru.

**Comment by troy-kuhns** [comment]
We have noticed all of the above as well and Fleet's lack of support for spaces has been a deal breaker for us transitioning. To add to this, even adding namespaces to the integrations does not resolve this problem since integration names still need to be unique across namespaces (which from a programing perspective makes me question the use of a namespace). It would be nice if this issue saw more traction, because in my opinion it's a gigantic oversight in terms of bringing multi-tenancy to Elastic.

**Comment by matthiasledergerber** [comment]
We do have exactly the same issue. Furthermore, we do have issues where the Integrations suddenly start to be installed to random spaces but the referenced data view still exists in another space that leads to broken dashboards. It seems as the current solution or workaround is not to use spaces at all when using integration. We do have a large elastic-stack running in an enterprise environment but this lack of consistency with the integrations and spaces make the out-of-the-box dashboards useless. Furthermore they randomly start to reinstall in different spaces confusing the users and there is no way except manually deleting the integration.

We currently have integrations that are in a broken state and require the removal of all objects in the spaces. But even then the reinstall routine somehow automatically starts to add objects everywhere resulting in multiple duplicated data views that initially work but break after restarting kibana. very funky.

I've already posted this issue in the forum, hopefully creating some traction because it baffels me that this doesnt' work correctly. I'm sure many others have this issue too.

[Elastic Forum Kibana Integration issue](https://discuss.elastic.co/t/kibana-fleet-high-cpu-load-on-elasticsearch-when-adding-integrations/336729/2)

**Comment by zez3** [comment]
Same issue here. I have a case with support but it is a pain to try to solve this

**Comment by yakhatape** [comment]
Hi @colin-stubbs , fun fact or not but im facing literraly to the same issue.
I started to work on elk v8 with the dream to realise an isolation between index data by patterns and space with the flexibility of the fleet management but it was stopped Really quickly with the assets management between spaces. With looks like pain in the ass....

This multi tenancy concept from elastic looks like not complete and will not be concedered by dev team when i check the date of your report...

I hope elastic will do something on this part because it can be a big plus for the product of its works properly...

I think we will go back with classic beats without any integration between spaces .. :(

**Comment by yakhatape** [comment]
@zez3 did you got an answer from support ?

**Comment by zez3** [comment]
@yakhatape 
I did. @nimarezainia can you say more about planning or if there is active work on this?

**Comment by nimarezainia** [comment]
@zez3 as noted Fleet's space awareness is one of our priorities, details of which are currently being worked on including how these assets will be constrained to a specific space.

**Comment by nimarezainia** [comment]
> Zero answer. I imagine they're happy with the complexity challenge as it'll lead to more independent deployments and more revenue.
> 
> It's still a workable situation if you're multi-tenanting things, but you definitely shouldn't give a customer any visibility of fleet.

@colin-stubbs I apologize for the difficulty you are having here. Please don't assume there's any malice for the sake of revenue generation. We care deeply about how the product is being used and endeavor to address all these popular deployment use cases. Supporting spaces is a priority for us so we hope to provide a timeline for its support. thank you.

**Comment by colin-stubbs** [comment]
No stress on my part anymore, I'm no longer operating a business that needs it.

The "we care deeply" part is hard to see... engagement here on GitHub is hit and miss, as you can see above with a bunch of people echo'ing that they have hit the same kind of issues, with zero acknowledgement or response from an Elastic employee until today.

Even when paying for a commercial subscription via Elastic Cloud, interacting with Elastic Cloud support gets very poor responses to issues like this. And that support team seems to be quite disconnected from Elastic development from my experience trying to log issues with them.

Though it's not really a surprise, as that support team has responsibility and a focus on keeping Elastic Cloud deployments running.

So there's a huuugggeee gap that needs be filled in terms of providing paying customers the ability to actually log problems or feature requests in order to get those issues onto a relevant Elastic teams radar.

**Comment by yakhatape** [comment]
> @zez3 as noted Fleet's space awareness is one of our priorities, details of which are currently being worked on including how these assets will be constrained to a specific space.

@nimarezainia do you have any target date about it ? Weeks , months , years ? Or maybe a "roadmap" about it to share with us?

**Comment by kpollich** [comment]
Hi all. I'm the tech lead on Fleet + Integrations, and I'd like to provide as much of an update as I can here.

First of all, I want to say thank you for the diligent reporting and detailed comments explaining Fleet's various deficiencies, quirks, and undocumented behaviors around multi-space support in Kibana. It's encouraging to see users providing detailed, well thought out feedback in public. We appreciate it tremendously. 

Second, I apologize for the lack of movement and communication on this issue in particular. Fleet's multi-space story has been discussed at length internally several times before, and there have been upstream projects in Kibana that we on the Fleet/Integrations team have been waiting on to begin making meaningful improvements in this area. Of course there will always be competing priorities as well (see our new [serverless offering](https://www.elastic.co/blog/elastic-serverless-architecture), Kafka support for Fleet-managed Agents, [secrets support for integration variables](https://github.com/elastic/kibana/issues/154715), and many more features we've delivered in the last year). Still, there's no excuse for ghosting our paying customers on this issue. 

We absolutely should've communicated these upstream blockers and changes in priority better to our community members and users. Why do we have public GitHub repos if not to provide transparency and direct insight into Elastic's engineering decisions and processes? Treating improvements in this area as blocked without transparently and explicitly relaying that decision to our open source community was a miss on our part. 

The largest blocker to a better multi-space story for Fleet: Kibana needs to have first-class support for "shareable" assets across spaces. "Assets" are things like dashboards, saved searches, visualizations, data views, tags etc. Essentially anything an integration can ship that gets created and referenced in Kibana when a given integration is installed. This is a very large undertaking, as historically the main means for sharing asset content across spaces has been basic duplication. Copying and pasting "managed" assets like those Fleet manages for integrations is a bad experience, though, because those copied assets are left behind when packages are altered or upgraded. 

The "shareable" content project is a massive undertaking that spans multiple teams and many core pieces of Kibana. Dashboards are essentially collections of many other asset types, and each of those asset types needs to be updated to be shareable using new sharing APIs and paradigms. The Kibana presentation team is hard at work on this, and you can track their progress on this large meta issue: https://github.com/elastic/kibana/issues/167901.

This is not to say that the Fleet team is entirely blocked on making incremental improvements for users making heavy use of spaces today. We can improve docs and fix bugs to mitigate the headaches we see with multi space environments today. The detailed reports in this issue and in support cases that we've seen in the last few months will inform those fixes, and we've already been making an effort to start patching up these smaller issues while we wait on the larger efforts above, e.g.

- https://github.com/elastic/kibana/pull/170789
- https://github.com/elastic/kibana/pull/168097
- https://github.com/elastic/kibana/pull/164469
- https://github.com/elastic/kibana/pull/164456
- https://github.com/elastic/ingest-docs/issues/708 (Docs improvement merged today!)

Moving forward, we will use this issue as a communication hub for improvements and movement on Fleet's multi-space story. Additionally, I will take it on as an action item to coalesce the detailed feedback and issues documented here and in several recent support cases into actionable issues and similar bugfix style PR's. There's no sense sitting on our hands waiting for the shareable dashboard initiative to complete when we can make meaningful improvements for our multi-space users here and now. Look here for comments like this one where we collect recent bug fixes, new issues, and docs updates related to multi-space behavior in Fleet in the coming months.

Now for a few things from the comments above that I can _not_ answer, however:

> @nimarezainia do you have any target date about it ? Weeks , months , years ? Or maybe a "roadmap" about it to share with us?

The product folks are hard at work on our product roadmap for the upcoming calendar year. As we establish requirements and specifics in the coming few weeks for the multi-space efforts related to Fleet we'll know more. I think a safe bet would be "definitely not weeks, and probably not years" though I know that's quite noncommittal. I'll let Nima give a better answer when he's able. 

> Even when paying for a commercial subscription via Elastic Cloud, interacting with Elastic Cloud support gets very poor responses to issues like this. And that support team seems to be quite disconnected from Elastic development from my experience trying to log issues with them.
> 
> Though it's not really a surprise, as that support team has responsibility and a focus on keeping Elastic Cloud deployments running.
> 
> So there's a huuugggeee gap that needs be filled in terms of providing paying customers the ability to actually log problems or feature requests in order to get those issues onto a relevant Elastic teams radar.

This is valuable feedback @colin-stubbs, so thank you for that. I'd be curious to hear more specifics here if you're willing and I'm sure someone from the support team would be eager to hear as much feedback as you're willing to give as well. @lucabelluccini sorry to volunteer you, but you're typically my go-to support resource 😅 - if there's someone better on the support management side feel free to ping them here as well. I don't want this feedback to go unheard by the right people who can take action. 

Thank you all for patience and feedback along the way. I hope this has been helpful, and please feel free to reach out with any questions or further feedback.

**Comment by colin-stubbs** [comment]
Fantastic response @kpollich , thank you.

**Comment by lucabelluccini** [comment]
Hello 👋 
I was tracking this issue too - but thank you for looping me in @kpollich

We are aware Fleet Integrations and Kibana Spaces support is limited.
- https://github.com/elastic/ingest-docs/issues/708
- https://github.com/elastic/ingest-docs/issues/248
- https://github.com/elastic/ingest-docs/issues/324

We have few Known issues too in our support portal (https://support.elastic.co/knowledge/f93188c4, https://support.elastic.co/knowledge/8c2e1720, https://support.elastic.co/knowledge/41de32b8).

As @kpollich said, this feature is dependant on Kibana "platform" (https://github.com/elastic/kibana/issues/167901).

As members of support, we usually raise enhancement requests on customer behalf to the product managers when there's a gap in the product or to report the importance of a 

I would invite @colin-stubbs to open a support case so we can follow up on the issues and concerns with the Support team to see how we can improve. Thank you!

**Comment by mgevans-5** [comment]
@kpollich thanks for the reply on this.  

If I'm reading this correctly most of the response is around sharing objects between spaces in a way to eliminate duplicate assets and enable update-one-consume-many.  

Can you please confirm if anything is also in the works (sorry if i missed it) to make Fleet and Spaces more tenant aware?  If I'm reading the original issue statement correctly the first part talks about the lack of separation between spaces which makes securing tenants difficult if not impossible.  We, too, were stymied by this challenge and it prevented us from rolling out Fleet to a many-dozen team implementation because we could not have them stepping on each others policies, let alone have access to privileged account information.

So for a multi-tenant environment (even internal to a company) is there progress anticipated in that through the other PRs that you've mentioned?  Or is this still a mostly unaddressed issue?

Thanks for your time and engagement to this group!

**Comment by kpollich** [comment]
Hey @mgevans-5, thanks for reaching out. 

> If I'm reading this correctly most of the response is around sharing objects between spaces in a way to eliminate duplicate assets and enable update-one-consume-many.
> 
> Can you please confirm if anything is also in the works (sorry if i missed it) to make Fleet and Spaces more tenant aware? If I'm reading the original issue statement correctly the first part talks about the lack of separation between spaces which makes securing tenants difficult if not impossible. We, too, were stymied by this challenge and it prevented us from rolling out Fleet to a many-dozen team implementation because we could not have them stepping on each others policies, let alone have access to privileged account information.

You are reading the above correctly, I was broadly speaking about sharing objects between spaces. 

As far as making Fleet more tenant-aware or supportive of multi-tenant environments: yes we have a large project in the works for achieving this within Fleet. I can't share much, as the project is still being defined internally, but it is something we've seen from many many customers (especially security use cases) since Fleet's initial release. 

I'll summarize some of what we have scoped out internally so far just for transparency:
- Allow for some level of granular permission/privileges on Fleet/Integrations data
- Map agent policies to a specific space
- Display agent policies and their enrolled agents only in the space to which they belong
- Only allow modification of agent policies for users who have the appropriate permissions within a space
- Allow multiple integration policies from multiple teams to be deployed to the same agent policy

Something I'll note to avoid confusion: "agent policies" as a data model are parents of the "integration policy" model that contains a given "instance" of an integration. So, broadly speaking, access control related to agent policies will also apply to integration policies as they're child objects of a given agent policy. 

There's more to be defined, but I think these broad requirements touch on the pain points we see in this thread. Once we have some public issues with the actual committed scope and implementation details for the multi-space effort we'll make sure to link them up here. At that point it will probably make sense to close this issue and move over to the public tracking issue for that as well. 

---

For now, I've spent some time gathering common issues and pain point related to Fleet and multi-space installations both from public issues and internal support cases. I got a very rough start on a tracking issue for incremental improvements we can make to ease those pain points while we define the broader multi-space effort: https://github.com/elastic/kibana/issues/172964. I've pulled a few of the issues there from the discussion here, but if there are any others we're happy to discuss them as well. As always, minimal reproductions are greatly appreciated (the description here is a perfect example).

**Comment by nimarezainia** [comment]
@mgevans-5 just curious whether adhering to spaces n Fleet and Integrations would be satisfactory for your use case? Are you expecting spaces, as a segregation of data concept, to also apply to your data? for example at the datastream level?

**Comment by mgevans-5** [comment]
Hi @nimarezainia 
We have interest in full segregation / tenancy between spaces.  This is for data as well as implementation components.  Our preferred model is to enable fast consumer implementation (also minimal centralized team) through secured access to their configurations in addition to basic data security.  While we've worked with the current data security-through-roles it would be easier for us to implement with data security-through-spaces (and also roles since individuals may have more/less access).

Whether the datastreams need to be space specific we're not concerned at this point.  The metadata filters have worked to date - however we've been looking at altering the index/stream configuration now that a proliferation and larger number of indexes/streams is less of a threat to the cluster health and memory consumption.  This may allow us to create space specific datastreams during ingest instead of relying on field filtering.

**Comment by nimarezainia** [comment]
thank you @mgevans-5. Would it be acceptable if all integrations that get installed, in context of a space, inherit the the space name as their Namespace and that way you get segregation at the datastream level?

for example, you are in space "coke" and install nginx. We fix the namespace in the nginx integration to "coke" and therefore your access logs would be at  `logs-nginx.access.coke`  (as oppose to default).  Would this be acceptable?

What @kpollich eloquently described above is about Fleet becoming space aware to ease some of the operational headaches we encounter. It doesn't apply to all the elements within the Elastic stack at this point. I am curious if this step (making Fleet/integrations space aware) would be useful for your deployment?

**Comment by colin-stubbs** [comment]
Hi all,

From my perspective data segregation has never really been an issue, Elasticsearch roles and index patterns and even field based limitations are sufficient.

It would be nice if this was more intuitive as part of space based configuration or role definition, but the capability is there.

For each instance of a integration *policy* the namespace is configurable, which results in a unique datastream for each *space*.

![Screenshot 2023-12-14 at 12 15 30 pm](https://github.com/elastic/integrations/assets/3059577/32b8c98d-d611-4834-b9f2-ee4566023b41)

For instance, I have "default" above, but that could also be "customerx".

![Screenshot 2023-12-14 at 12 15 53 pm](https://github.com/elastic/integrations/assets/3059577/e5c165b4-d57e-4a4f-9330-9b0401484310)

So, instead of "-default" at the end, if I configured the integration with "customerx" I would get datastreams like the above but with "-customerx" on the end.

You can use role based index patterning to restrict you tenant aka. space users to only seeing data in indices/datastreams that match "-customerx", which will result in what you want.

e.g. a really basic example, more complexity is actually required for production, I just can't remember the full details of what it was, but you get the idea hopefully.

![Screenshot 2023-12-14 at 12 32 26 pm](https://github.com/elastic/integrations/assets/3059577/3e683354-cb39-45b2-9b8a-305595e60619)

TL;DR - If you treat space as exactly equivalent to tenant, the you have data segregation and sufficient RBAC to control access to it.

The issue, has, and still is as per my original description, e.g. that Kibana doesn't understand or enforce RBAC to Kibana based configuration elements in the same way that Elasticsearch does to indices.

e.g. the issues again,

1. Fleet has effectively ZERO Role Based Access Control.

This is the main drama that people are facing and where the biggest room for improvement exists.

If you assign Fleet access in a Kibana role, it has to be to "All Spaces", there is no granular per-space RBAC that can be applied.

This means that if someone in "customerx" gets assigns Kibana privileges to use Fleet, they CANNOT be restricted to seeing installed integrations and configuration integration policies in "customerx" only, they will get everything.

They *will* wind up seeing everything in "default", "customerx" and "customery".

The "Integrations" option in the roles doesn't seem to do anything at present.

It's just this Fleet option that controls *everything*,

![Screenshot 2023-12-14 at 12 34 23 pm](https://github.com/elastic/integrations/assets/3059577/deb09f29-cd2f-4f88-91de-7be1cf15656b)

This also means a "customerx" user, who should never see or know about other spaces, will be prompted to select between "default", "customerx" and "customery" space when they login... which is at the very least less than desirable.

This also means that a "customerx" user can then open an integration policy for "customery", and extract details from it, including and not limited to secrets such as API keys.

e.g. This "customerx" user, should not be seeing this integration policy at all.. but they can, which means they can get to the JWT.

![Screenshot 2023-12-14 at 1 21 26 pm](https://github.com/elastic/integrations/assets/3059577/8ec99a81-3091-493d-b741-3e21044949c3)

The reason someone want to be able to give customers the ability to see and interact with their own integrations and integration policies, is from my experience, two-fold:

1. In read-only type use cases, it allows the customer to have confidence in what's configured by being able to see it and see how it's configured. They may even want to audit it because reasons.

2. In read-write type use cases, the customer staff may need to have the ability to at least modify existing integration policies to support maintenance activity such as API key rollovers, or minor changes such as changing tags or configuration data that gets attached to documents processed by the integration, e.g. mapping of serial numbers to hostnames etc. The customer's staff may also want to be able to perform useful functions like endpoint isolation if Elastic Endpoint Security is in use, or perform queries using the osquery integration, etc etc.

At the moment there's zero granularity to this and it's non-obvious if not impossible to deal with it.

This is still the case in 8.11.1.

What's the issue?

Well... I suspect that it's mostly just that all of the Fleet related configuration, for all spaces, is simply in hidden system indices that make more granular based access kinda hard?

e.g. these is where fleet is actually configured and all of the integration policy config exists...

![Screenshot 2023-12-14 at 1 07 59 pm](https://github.com/elastic/integrations/assets/3059577/d385a42c-7677-4bb2-aa06-18e626269b0a)

Splitting integration policy configuration out into separate indices per space might be at least part of the solution so that index based patterning (or a similar approach behind the scenes in Kibana) can be applied?

----------------------------------------------------------------

2. Kibana gets confused about where the integration assets are installed, e.g. it treats them as only installed in the first space that the integration was installed via.

This means dashboards and other assets are only ever installed in the space the integration is *first* installed via, and for any other spaces where the integration is being "used" we have to manually copy them to other spaces.

This also means that updates to official integration dashboards and other saved objects also don't automatically get applied to anywhere else except the space that the original assets are in.

See previous description at the start of this issue.

Nothing's changed with this as of 8.11.1 as best I can tell.

The ability to "install" the integration to multiple spaces and have assets auto-update in them is what people want here.

It might be more useful to think of integrations as being "installed" to the deployment, but "used" or "not used" in a space.

e.g. this is not useful.

![Screenshot 2023-12-14 at 12 59 43 pm](https://github.com/elastic/integrations/assets/3059577/2c554ad0-c51f-4fbd-a289-512837707d4b)

e.g. I can do a "copy to space" individually on EACH AND EVERY COMPONENT, BUT THIS IS NOT USEFUL.

![Screenshot 2023-12-14 at 1 52 39 pm](https://github.com/elastic/integrations/assets/3059577/1310a400-462c-4441-a86b-c250446486e5)

The first one will copy fine,

![Screenshot 2023-12-14 at 1 55 15 pm](https://github.com/elastic/integrations/assets/3059577/973448e2-db78-403e-9a22-3dfd052b2075)

But now there's more conflicts and I have to do more clicky clicky.

![Screenshot 2023-12-14 at 1 55 48 pm](https://github.com/elastic/integrations/assets/3059577/75f526a6-d301-46e3-be0c-e4de35a03041)

e.g. manual export and import required to get these into space "customerx" is the least frustrating way,

![Screenshot 2023-12-14 at 1 52 54 pm](https://github.com/elastic/integrations/assets/3059577/e4af1552-f18d-4561-87ae-6e73513aa878)

![Screenshot 2023-12-14 at 1 58 22 pm](https://github.com/elastic/integrations/assets/3059577/9eef4e74-8e73-455c-968c-18d8d8277049)

![Screenshot 2023-12-14 at 1 58 41 pm](https://github.com/elastic/integrations/assets/3059577/23d58ed9-4e27-473d-a94a-b00190641eba)

But STILL, from Fleet's perspective, the integration is NOT INSTALLED IN "customerx".

![Screenshot 2023-12-14 at 1 59 16 pm](https://github.com/elastic/integrations/assets/3059577/10a98c43-b973-48b9-9bb5-a050e98502e0)

Hence those copied saved objects will never be updated as part of an integration update.

**Comment by zez3** [comment]
I actually use the before mentioned "tenants" example (in my case are internal/external teams/institutes) space to role(based on namespace) RBAC for my productive environment. I also limit the access via `Grant read privileges to specific documents`+`Granted documents query`
So, that works. 

I also tag all of my Agents in Fleet with a specific tenant tag. 
![image](https://github.com/elastic/integrations/assets/22619253/f37f473e-60e2-4eff-85d3-5cf2426a4fb6)

I was thinking that we could use that tag and allow a similar RBAC in Fleet. This could fix the issue of tenants own Agents. I mean for read-only status, not full access.
The same tag can/should be used when the Agent is enrolled. That should help my sysadmin 
I am not sure if this Fleet tag was ever though of doing more than displaying a group of agents.

I am also hitting the integration issue:

>The ability to "install" the integration to multiple spaces and have assets auto-update in them is what people want here.
>It might be more useful to think of integrations as being "installed" to the deployment, but "used" or "not used" in a space.

Indeed, this is what we will eventually need.

Now to gain/give write permissions for tenants...
If I where to give access to my tenant to create own Fleet policies(with own integrations) then they will need to be limited only to the assigned namespace. Otherwise the Role has to be changed each time they create a new policy with a new namespace
Could we perhaps do this via the default namespace inherited from the selected Agent policy?
I was thinking that the Elastic admins could create such policies in advance and the tenant still be limited base on the namespaces/index_names in roles.
Another issue is with DataViews in Kibana. They can only be created after the index is created. I would like to be able to create then before. Not sure if there is an issue for this already.

**Comment by colin-stubbs** [comment]
@zez3  I'm 99.999% certain you can work around the data view chicken/egg problem by just creating the data view via REST API. The Kibana webUI enforces the existence of matching indices/datastreams, however the REST API at /s/{{space_id}}/api/data_views/data_view does not.

Doco here: https://www.elastic.co/guide/en/kibana/current/data-views-api-create.html

FYI - The way I was automating onboarding of new "tenants" was via a Postman ( https://www.postman.com/ ) collection that created all of the necessary roles, users, Fleet integration policies etc etc etc, including data views and adjusting Kibana advanced setting customisation. This all happened prior to any indices/datastreams being created in any way.

It'll remove most of the risk of human error and lack of consistency if you approach it that way.

**Comment by kpollich** [comment]
Hi folks, I appreciate the extremely detailed discussions here, and I assure you we have no plans to ignore any of this - in fact the detailed walkthroughs and screenshots in the comments above are greatly informing our definition of the ongoing multi-space Fleet project we're looking to deliver in the near future. That being said, I am closing this issue in favor of https://github.com/elastic/kibana/issues/175831 as the actual issues described here are related to Kibana, not the integrations repository. In an effort to place the discussion as close as possible to the code to which it relates, I've created the issue above as a centralized hub for future discussion around multi-space usage of Fleet/Integrations. I've also provided extensive context, and will provide continual updates as we make progress on first-class multi-space support in the coming months. Thanks for your continued patience and thanks again for so much productive discussion on this issue.
